libraries/stdlib/jvm/src/kotlin/io/files/FileTreeWalk.kt

```diff
@@ -108,11 +137,40 @@ public class FileTreeWalk private constructor(
                 } else {
                     // Proceed to a sub-directory
                     state.push(directoryState(file))
-                    return gotoNext()
+                    return gotoNextDepthFirst()
                 }
             }
         }
 
+        /**
+         * Walking with BFS algorithm
+         */
+        private fun gotoNextBreadthFirst(): File? {
+            val walkState = state.poll() ?: return null
+            val file = walkState.root
+            if (currentDepth >= maxDepth) {
+                return file
+            }
+
+            // To process maximum depth
+            val nameCount = file.toComponents().size
```

Comment:
I believe, the next depth level could be easily determined without splitting each file to components: instead another queue for the next level should be introduced. When "this level" queue becomes empty, it is swapped with "the next level" queue and then processing continues with that next level queue.

Reply:
Or perhaps some other technique could be used to determine where the next level of depth is reached even in the single queue. For example by incrementing a next-level counter when enqueueing items and decrementing this-level counter when dequeueing items.

Reply:
Ok, I need to think about it and I will choose better technique.

Reply:
@ilya-g What do you think about store current depth of node in `kotlin.io.FileTreeWalk.WalkState`? This class is private (and all his childrens), so public API will not be affected.

Reply:
@aivinog1 That's a viable solution too, though it could slightly increase memory consumption.

Btw, it might be easier to not try to shoehorn BFS into the existing WalkState classes, but rather write BFS algorithm separately with the coroutine-based sequence and iterator builders.

E.g. in `FileTreeWalk.iterator` function:
```
fun iterator() = if (direction == BFS) bfsIterator() else FileTreeWalkIterator()

private fun bfsIterator(): Iterator<File> = iterator {
   // write BFS iteration algorithm imperatively here, 
   // use yield and yieldAll functions to returns files to sequence consumer
}
```
If you find this easier, you can go this way. But if not, it's not a problem: you can continue with the current approach and we'll investigate this refactoring separately.